server_status.py

```diff
@@ -0,0 +1,253 @@
+import subprocess
+import os
+import time
+import mraa
+import inspect
+import sys
+
+os.chdir('/home/root/gpio')
+print "*** START server_status.py ***"
+
+# FUNCTIONS
+
+# Returns current line number
+def lineno(): 
+    return inspect.currentframe().f_back.f_lineno
+
+def red_on():
+  global red
+  if not red:
+    print "red on"
+    red = True
+    red_led.write(1)
+
+def red_off():
+  global red
+  if red:
+    print "red off"
+    red = False
+    red_led.write(0)
+
+def green_on():
+  global green
+  if not green:
+    print "green on"
+    green = True
+    green_led.write(1)
+
+def green_off():
+  global green
+  if green:
+    print "green off"
+    green = False
+    green_led.write(0)
+
+def blue_on():
+  global blue
+  if not blue:
+    blue = True
+    print "blue on"
+    blue_led.write(1)
+
+def blue_off():
+  global blue
+  if blue:
+    print "blue off"
+    blue = False
+    blue_led.write(0)
+
+def reset():
+  global red
+  global green
+  global blue
+  if red:
+    red_off()
+  if green:
+    green_off()
+  if blue:
+    blue_off()
+
+
+def get_openmrs_internal_status():
+  with open('/home/root/debian/home/buendia/server_status.txt', "r") as status_file:
+    status = status_file.read()
+    return int(status)
+
+def report(n):
+  print time.strftime('%X') + ' - line ' + str(n+1)
+
+def openmrs_internal_status():
+  global openmrs_i
+  openmrs_i_new = get_openmrs_internal_status()
+  if not openmrs_i_new == openmrs_i:
+    reset()
+    openmrs_i = openmrs_i_new
+  # TASK LED's according to status
+  if openmrs_i == 0:       # down - GREEN on
+    report(lineno())
+    reset()
+    report(lineno())
+    green_on()
+  elif openmrs_i == 1:     # normal use - BLUE on
+    report(lineno())
+    reset()
+    report(lineno())
+    blue_on()
+  elif openmrs_i == 2:     # activity  - blink BLUE on/off
+    if blue:
+      report(lineno())
+      blue_off()
+    else:
+      report(lineno())
+      blue_on()
+  elif openmrs_i == 3:     # backing-up - blink BLUE/GREEN
+    if blue:
+      report(lineno())
+      blue_off()
+      report(lineno())
+      green_on()
+    else:
+      report(lineno())
+      green_off()
+      report(lineno())
+      blue_on()
+  elif openmrs_i == 4:     # back-up failed - blink BLUE/RED
+    if blue:
+      report(lineno())
+      blue_off()
+      report(lineno())
+      red_on()
+    else:
+      report(lineno())
+      red_off()
+      report(lineno())
+      blue_on()
+
+# classify server repsonses - i.e. interpret top line of 'curl -Is <url>'
+def classify_status(sts):
+  if sts == '':               # no response
+    return 0
+  elif sts.find('200') > -1:  # positive response
+    return 1
+  else: # other response (e.g. redirected) [Might have to debug this..]
+    return 2
+
+
+# INITIALISE SETTINGS
+
+red = False
+red_led = mraa.Gpio(31)     # 'GP44'
+red_led.dir(mraa.DIR_OUT)
```

Comment:
If we really need to record the state of each LED, then encapsulate it in a class.

I don't see why we need to record the on/off state though.  I assume sending an "on" command to an LED that's already on is harmless and takes negligible time.  For simplicity can we reduce the entire API to:

    set_led(identifier, state)

    identifier: RED, GREEN, or BLUE
    state: True or False